File: lb-processor/src/main/java/com/github/rooneyandshadows/lightbulb/annotation_processors/utils/ElementUtils.java
package com.github.rooneyandshadows.lightbulb.annotation_processors.utils;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;

public class ElementUtils {

    public static TypeName getTypeOfFieldElement(Element element) {
        return ClassName.get(element.asType());
    }

    public static boolean canBeInstantiated(Element classElement) {
        return classElement.getKind() == ElementKind.CLASS && !classElement.getModifiers().contains(Modifier.ABSTRACT);
    }

    public static String getPackage(Elements elements, Element element) {
        return elements.getPackageOf(element)
                .getQualifiedName()
                .toString();
    }

    public static TypeElement getSuperType(TypeElement element) {
        TypeMirror superClassTypeMirror = element.getSuperclass();
        return (TypeElement) ((DeclaredType) superClassTypeMirror).asElement();
    }

    public static String getSimpleName(Element element) {
        return element.getSimpleName().toString();
    }

    public static String getFullClassName(Elements elements, Element element) {
        String classPackage = getPackage(elements, element);
        String classSimpleName = element.getSimpleName().toString();
        return classPackage.concat(".").concat(classSimpleName);
    }

    public static String scanForSetter(Element classElement, String fieldName) {
        String setterName = "set".concat(capitalizeFirstLetter(fieldName));
        return methodExists(classElement, setterName) ? setterName : null;
    }

    public static String scanForGetter(Element classElement, String fieldName) {
        String getterName = "get".concat(capitalizeFirstLetter(fieldName));
        return methodExists(classElement, getterName) ? getterName : null;
    }

    private static boolean methodExists(Element classElement, String methodName) {
        if (classElement.getKind() != ElementKind.CLASS) return false;
        return classElement.getEnclosedElements().stream()
                .anyMatch(target -> {
                    String targetName = target.getSimpleName().toString();
                    boolean take = target.getKind() == ElementKind.METHOD;
                    take &= targetName.equals(methodName);
                    return take;
                });
    }

    private static String capitalizeFirstLetter(String target) {
        return target.substring(0, 1).toUpperCase() + target.substring(1);
    }
}

File: src/main/java/com/github/rooneyandshadows/lightbulb/annotation_processors/generator/utils/ElementUtils.kt
package com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils

import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.booleanPrimType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.booleanType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.doublePrimType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.doubleType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.floatPrimType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.floatType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.intPrimType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.intType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.longPrimType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.longType
import com.github.rooneyandshadows.lightbulb.annotation_processors.generator.utils.TypeUtils.stringType
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.asTypeName
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.Modifier
import javax.lang.model.util.Elements

internal object ElementUtils {
    @JvmStatic
    fun getPackage(elements: Elements, element: Element?): String {
        return elements.getPackageOf(element)
            .qualifiedName
            .toString()
    }

    @JvmStatic
    fun getFullClassName(elements: Elements, element: Element): String {
        val classPackage = getPackage(elements, element)
        val classSimpleName = element.simpleName.toString()
        return "$classPackage.$classSimpleName"
    }

    @JvmStatic
    fun getTypeOfFieldElement(element: Element): TypeName {
        var typeName = element.asType().asTypeName()
        when (typeName.toString()) {
            stringType -> {
                typeName = ClassName("kotlin", "String")
            }

            intType, intPrimType -> {
                typeName = ClassName("kotlin", "Int")
            }

            booleanType, booleanPrimType -> {
                typeName = ClassName("kotlin", "Boolean")
            }

            floatType, floatPrimType -> {
                typeName = ClassName("kotlin", "Float")
            }

            longType, longPrimType -> {
                typeName = ClassName("kotlin", "Long")
            }

            doubleType, doublePrimType -> {
                typeName = ClassName("kotlin", "Double")
            }

            else -> {}
        }
        return typeName
    }

    @JvmStatic
    fun canBeInstantiated(classElement: Element): Boolean {
        return classElement.kind == ElementKind.CLASS && !classElement.modifiers.contains(Modifier.ABSTRACT)
    }
}